rust/bridge/node/futures/src/promise.rs

```diff
@@ -0,0 +1,133 @@
+///     let trait_impl = TraitImpl::new(cx.queue(), js_info.root(&mut cx));
+///     promise(&mut cx, async move {
+///         let result = compute_result(trait_impl).await?;
+///         fulfill_promise(move |cx| Ok(cx.string(result)))
+///     })
+/// }
+/// ```
+pub fn promise<'a, V, F>(
+    cx: &mut FunctionContext<'a>,
+    future: impl Future<Output = Result<F, PersistentException>> + UnwindSafe + 'static,
+) -> JsResult<'a, JsObject>
+where
+    V: neon::types::Value,
+    F: for<'b> FnOnce(&mut TaskContext<'b>) -> JsResult<'b, V> + Send + UnwindSafe + 'static,
+{
+    let callbacks_object = cx.empty_object();
+    let save_promise_callbacks_fn = JsFunction::new(cx, save_promise_callbacks)?;
+    let bind_fn: Handle<JsFunction> = save_promise_callbacks_fn
+        .get(cx, "bind")?
+        .downcast_or_throw(cx)?;
+    let bound_save_promise_callbacks =
+        bind_fn.call(cx, save_promise_callbacks_fn, vec![callbacks_object])?;
+
+    let promise_ctor: Handle<JsFunction> = cx.global().get(cx, "Promise")?.downcast_or_throw(cx)?;
+    let promise = promise_ctor.construct(cx, vec![bound_save_promise_callbacks])?;
+
+    let callbacks_object_root = callbacks_object.root(cx);
+    let queue = cx.queue();
+
+    cx.run_future_on_queue(async move {
+        let result = future.catch_unwind().await;
+
+        queue.send(move |mut cx| -> NeonResult<()> {
+            let resolved_result = match result {
+                Ok(Ok(resolve)) => {
+                    // This AssertUnwindSafe is not technically safe.
+                    // If we get a panic downstream, it is entirely possible the JavaScript context won't be usable anymore.
+                    // However, that's no more unsafe than JsAsyncContext::with_context,
+                    // or for that matter Neon automatically catching panics by default.
+                    let mut cx = AssertUnwindSafe(&mut cx);
+                    catch_unwind(move || cx.try_catch(|cx| resolve(cx)))
+                }
+                Ok(Err(exception)) => Ok(Err(exception.into_inner(&mut cx))),
+                Err(panic) => Err(panic),
+            };
+            let folded_result = resolved_result.unwrap_or_else(|panic| {
+                Err(cx
+                    .error(format!("unexpected panic: {}", describe_any(&panic)))
+                    .expect("can create an Error")
+                    .upcast())
+            });
+
+            let (value, callback_name) = match folded_result {
+                Ok(value) => (value.upcast(), RESOLVE_SLOT),
+                Err(exception) => (exception, REJECT_SLOT),
+            };
+
+            let callbacks_object = callbacks_object_root.into_inner(&mut cx);
+            let promise_callback: Handle<JsFunction> = callbacks_object
+                .get(&mut cx, callback_name)?
+                .downcast_or_throw(&mut cx)?;
+            promise_callback.call(&mut cx, callbacks_object, vec![value])?;
+            Ok(())
```

Comment:
I think this chunk of code could be more clear.  Perhaps with something like this:

```
match result {
  Ok(Ok(resolve)) => {
    match catch_unwind(|| AssertUnwindSafe(cx).try_catch(resolve)) {
      Ok(value) => {
       js_call_method(callbacks_object, RESOLVE_SLOT, vec![value])
      }
      Err(panic) => {
       err = cx.error(format!("unexpected panic: {}", describe_any(panic))
       js_call_method(callbacks_object, REJECT_SLOT, vec![err])
      }
     }
    }
    Ok(Err(exception)) => {
     err = Err(exception.into_inner(cx)
     js_call_method(callbacks_object, REJECT_SLOT, vec![err])
    }
    Err(panic) => {
       err = cx.error(format!("unexpected panic: {}", describe_any(panic))
       js_call_method(callbacks_object, REJECT_SLOT, vec![err])
    }
   }
```
It's a little more duplicative, but I think it's much easier to understand.

Reply:
I felt like it was important to treat the panics uniformly, but I'll at least add some explicit types to make it clearer what's going on. I agree that the `callback_name` bit is too clever, though.